### DISCUSSION TITLE
What's the local dev story?

### DISCUSSION THREAD

**Discussion by devlife** [OrchardCMS/OrchardCore#13478] [Q&A] [answered]
Is there an official or supported way of developing locally with source code? Or more specifically, is there an easy way to go back and forth between using OC source code (presumably while learning or troubleshooting) and the nuget packages (for compilation performance, etc)?

**Comment by jtkech** [comment]
AFAIK There is no official way ;)

In fact there are many scenarios so it depends on the needs.

Yes, for building perf you can reference OC as packages from nuget/cloudsmith, then for example you can reference a custom module in dev as a project , in that case you can still debug OC code with `sourcelink` but less confortable. Then to test your module as a package, you can pack/push it to a package manager (or a local feed folder) and reference it.

Personnally I mainly work with the full source code for testing / debugging, then to see if it works in release / prod mode (e.g. with assets embedded in assemblies), most of the time it is sufficient to publish the app in a local folder and run it from there, and in this case you can still attach the process to the debugger.

But for a full customer application always better at the end to reference OC as packages, then you can add to the solution your own custom modules/libraries projects/packages/github-submodules.

**Reply by devlife** [reply]
> But for a full customer application always better at the end to reference OC as packages, then you can add to the solution your own custom modules/libraries projects/packages/github-submodules.

Can you explain this last bit? The vast majority of the work I'm doing is for customer themes and modules.

**Comment by jtkech** [comment, marked as answer]
> The vast majority of the work I'm doing is for customer themes and modules.

So not a full application, hmm here I would recommend to fork and use the full source code, and add references to your custom modules/themes from the web app or for example as we do in `OC.Application.Cms.Core.Targets`.

But at some point, as a final test for a given module/theme, pack it and reference it as a package.

**Reply by devlife** [reply]
I'll have to look into that. Thanks very much!!
